File: C++/C++/_128.cpp
#include "_128.h"

int _128::longestConsecutive(std::vector<int>& nums) {
	std::unordered_map<int, int> map;

	if (nums.size() == 0) return false;

	for (auto n : nums) {
		if (map.count(n) == 0) {
			map[n] = 1;
		}
	}
	int maxVal = 1;
	for (auto n : map) {
		int k = n.first - 1;
		while (map.count(k) != 0)
		{
			if (map[k] != 1) {
				n.second = map[k] + n.first - k;
				break;
			}
			n.second++;
			map[k] = map[k+1] - 1;
			k--;
		}
		maxVal = std::max(maxVal, n.second);
	}
	return maxVal;
}

File: C#/C#/_128.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    public class _128
    {
        public int LongestConsecutive(int[] nums)
        {
            Dictionary<int, int> dic = new Dictionary<int, int>();

            foreach (int n in nums)
            {
                if (dic.ContainsKey(n)) continue;
                dic.Add(n, 1);
            }

            int max = 0;
            foreach (int n in nums)
            {
                int k = n - 1;
                while (dic.ContainsKey(k))
                {
                    dic[n]++;
                    dic[k] = dic[k + 1] - 1;
                    k--;
                }
                max = Math.Max(max, dic[n]);
            }
            return max;
        }
    }
}
